[WIP] Adds CustomEntanglerLayers

Lucaman99 opened the pull request:
**Context:**

https://github.com/XanaduAI/pennylane/issues/680

**Description of the Change:**

**Benefits:**

Allows for more customizability of entangling layers.

**Possible Drawbacks:**

None that I can think of

**Related GitHub Issues:**

https://github.com/XanaduAI/pennylane/issues/680

Note: @josh146 I'm not sure if this is completely working with the new qml.wires.Wires class (specifically, non-integer labels for wires). I know that this is a fairly recent change to PennyLane, so are related bugs still in the process of being fixed?

josh146 commented:
> Note: @josh146 I'm not sure if this is completely working with the new qml.wires.Wires class (specifically, non-integer labels for wires). I know that this is a fairly recent change to PennyLane, so are related bugs still in the process of being fixed?

The introduction of the new `Wires` class is still ongoing, but looks like all the tests are so far passing!

Lucaman99 commented:
@josh146 I've been thinking about how we can go about taking weights as arguments for the parametrized two-qubit gates, and I was wondering: do you think it would make sense for the BasicEntengler to take two arguments, rotation_weights and coupling_weights for each of the rotation and coupling gates respectively?

josh146 commented:
> do you think it would make sense for the BasicEntengler to take two arguments, rotation_weights and coupling_weights for each of the rotation and coupling gates respectively?

@mariaschuld would be the best person to answer this question 🙂 The only blocker I can think of would be QNodeCollections, which currently assume the ansatz has a signature `qnode(params, **kwargs)`.

co9olguy commented on pennylane/templates/layers/basic_entangler.py:
```diff
@@ -23,12 +23,14 @@
 
 @template
-def BasicEntanglerLayers(weights, wires, rotation=None):
+def BasicEntanglerLayers(weights, wires, rotation=None, interactions=None):
```
I think it might make sense to leverage `broadcast` more directly, so that these changes proposed changes can be used outside of just `BasicEntanglerLayers`. 

What I'd suggest is the following:
- change the keyword arg here to "pattern" (same as `qml.broadcast`)
- add this new functionality (ability to arrange two-qubit gates according to a custom user-specified pattern) into `qml.broadcast`
- in this template, rely on that implementation

Lucaman99 replied:
Yeah, I was talking to Maria in the thread of the original issue, and she says that it makes more sense to just make an entirely new template, `EntanglerLayers`, so that the "basic-ness" of this template can be preserved.

mariaschuld replied:
Yes @co9olguy 's suggestion would make a lot of sense. 

Upon second thought, if the ``pattern`` argument has a default, a user would never have to touch it if the template is just used in its basic form...and since the entanglers do not take parameters it is still pretty simple to use.

@Lucaman99 if you prefer to extend this template, I'll be in support!

Lucaman99 replied:
@mariaschuld You're definitely the expert, so I'll defer to your judgment 🙂 . Which option do you think is best? I'm now thinking that your idea to make another `EntanglerLayers` template would be optimal, since it would be easier to have two separate arguments for the rotation parameters and the coupling parameters (if we are allowing for parametrized two-qubit operations).

mariaschuld replied:
I'll leave it up to you. If you go with ``EntanglerLayers`` you could also make the entangler gate a custom input.

What about calling it ``CustomEntanglerLayers``? That's just because we considered removing the ``Layers`` in all templates, and it would make sense in contrast to the ``Basic`` one..

Lucaman99 replied:
@mariaschuld That sounds good 👍

mariaschuld commented:
> > Note: @josh146 I'm not sure if this is completely working with the new qml.wires.Wires class (specifically, non-integer labels for wires). I know that this is a fairly recent change to PennyLane, so are related bugs still in the process of being fixed?
> 
> The introduction of the new `Wires` class is still ongoing, but looks like all the tests are so far passing!

I think templates that don't use the `Wires` class should still work fine...the advantage of using it is just the sanity checks it introduces.

Lucaman99 commented:
Hey @mariaschuld @josh146 - I reverted the `BasicEntanglerLayers` back its original form, and created a new template file called `CustomEntanglerLayers`. Basically, this template is like `BasicEntanglerLayers`, but allows the user to have 1. A custom pattern of coupling gates and 2. Parameters passed into the layers of coupling gates. 

I've played around with it a little bit, and I think I've got all of the edge cases accounted for (there are a lot!).

I still have to do quite a bit of work, including writing the comments, writing tests, and cleaning up the code, but I thought I would push what I have so far (this PR has also been converted to a draft).

mariaschuld commented on pennylane/templates/layers/custom_entangler.py:
```diff
@@ -31,7 +31,7 @@
+r"""Layers consisting of one-parameter single-qubit rotations on each qubit, followed by a sequence of 
     parametrized double-qubit gates
 
     The placement of double-qubit gates on the circuit is determined by a user-passed
```
This comment is for the rest of the docstring...I think you need to change the note (which is only true for the specific "ring" pattern that ``BasicEntanglers`` uses), and the figure?

Also, I would add an explanation for ``pattern``, saying how it works to customize the entangling gates?

mariaschuld commented on pennylane/templates/layers/custom_entangler.py:
```diff
@@ -212,31 +212,31 @@ def circuit(weights):
+                msg="'coupling_weights' must be of shape {}; got {}" "".format(expected_shape, get_shape(coupling_weights)),
+                )
+
+        elif type(pattern) == list:
```
I would use ``isinstance``, that is best practice because it also catches types that the object inherits from, if i am not mistaken.

mariaschuld commented on pennylane/templates/layers/custom_entangler.py:
```diff
@@ -249,4 +249,4 @@ def circuit(weights):
         else:
-            broadcast(unitary=coupling, pattern=pattern, wires=wires)
\ No newline at end of file
+            broadcast(unitary=coupling, pattern=pattern, wires=wires)
```
```suggestion
            broadcast(unitary=coupling, pattern=pattern, wires=wires)
            
```

Python standard to add a blank line :)

mariaschuld commented on pennylane/templates/layers/basic_entangler.py:
```diff
@@ -126,4 +126,4 @@ def circuit(weights):
         broadcast(unitary=rotation, pattern="single", wires=wires, parameters=weights[layer])
-        broadcast(unitary=CNOT, pattern="ring", wires=wires)
\ No newline at end of file
+        broadcast(unitary=CNOT, pattern="ring", wires=wires)
```
```suggestion
        broadcast(unitary=CNOT, pattern="ring", wires=wires)
        
```

Not sure what this change is about? Did someone delete the new line at the end?

Lucaman99 replied:
Yeah, that was my fault, I was checking to see that all the changes from the earlier version of this PR were reverted and I accidentally deleted the end line.

co9olguy commented on .github/CHANGELOG.md:
```diff
@@ -141,6 +141,11 @@
     The ability to pass `argnum` has been retained for backwards compatibility, and
     if present the old behaviour persists.
 
+* The `BasicEntanglerLayers` template now has an argument, `interactions`, which allows a custom 
+  braodcasting pattern of CNOT gates to be used within each layer, with "ring" being 
+  the default.
+  [(#681)](https://github.com/XanaduAI/pennylane/pull/681)
+
```
This no longer seems to be the case

co9olguy commented on pennylane/templates/layers/custom_entangler.py:
```diff
@@ -0,0 +1,262 @@
+
+@template
+def CustomEntanglerLayers(
+    rotation_weights, wires, rotation=None, coupling=None, coupling_weights=None, pattern=None
```
Before merging, it's good to think about the ordering of the keyword arguments.

Some things to keep in mind:
- signatures should mirror other similar functions as much as possible
- the arguments that the user would likely use most should be the earliest in the signature (since python lets you specify keyword args positionally sometimes)

Lucaman99 replied:
I'm thinking that it should probably be `pattern`, `rotation`, `coupling`, `coupling_weights`, what did you have in mind?

co9olguy commented on pennylane/templates/layers/custom_entangler.py:
```diff
@@ -0,0 +1,262 @@
+        "all_to_all": 0 if len(wires) in [0, 1] else len(wires) * (len(wires) - 1) // 2,
+    }
+
+    OPTIONS = ["single", "double", "double_odd", "chain", "ring", "pyramid", "all_to_all"]
```
Does it make more sense to import options from the broadcast function? That way it's more maintainable

Lucaman99 replied:
Yeah, this definitely makes more sense, although the OPTIONS and n_parameters variables are contained in the broadcast method, so is it possible to access them from importing braodcast.py?

co9olguy replied:
yep, that's what I had in mind

if the are inside the method, it's relatively simple to move them out to the larger namespace

Lucaman99 replied:
Ok, done!

Lucaman99 replied:
I added a new method to braodcast.py that returns the object of param numbers for each pattern (for an inputted number of wires), which I use in custom_entangler.py

co9olguy replied:
Just as a quick glance, that looks suitable, but I'll defer to @josh146 and @mariaschuld for the final code review

Lucaman99 replied:
@co9olguy Sounds good, thanks!

co9olguy replied:
lol do you have these notifications hard-wired into your cerebral cortex? you're always so fast

Lucaman99 replied:
Surprisingly no. Although it's pre-historic compared to Internet implanted in the brain, email notifications work pretty well 😅

co9olguy replied:
I had to turn those off a long-time ago :grinning:, way too many to sort through

co9olguy commented on pennylane/templates/layers/custom_entangler.py:
```diff
@@ -0,0 +1,262 @@
+            a Wires object.
+        rotation (pennylane.ops.Operation): one-parameter single-qubit gate to use,
+                                            if ``None``, :class:`~pennylane.ops.RX` is used as default
+        pattern (?????): A keyword that determined how the double-qubit gates will be placed on the circuit.`pattern='ring'` is used
```
Maybe could connect to the broadcast function here, so users know what the possible options are (and more maintainable if you link rather than reproduce the patterns here)

Lucaman99 commented:
Hey @mariaschuld @co9olguy - I'm pretty much done polishing this PR, so I just requested re-review! A lot has changed since you last looked at it:

- The CustomEntanglerLayers template file is added and is now cleaned up/has a docstring.
- There are now extensive tests for the CustomEntanglerLayer template
- Added functions in [broadcast.py](http://broadcast.py/) that return objects with the number of two-qubit gates in each pattern, and wire sequences for each pattern. This is used in the CustomEntanglerLayer, and will hopefully be useful for future maintainability, as new patterns can simply be added to [broadcast.py](http://braodcast.py/), and then properties of these patterns can be accessed from any other file (by calling these functions).
- Added two new methods in [init.py](http://init.py/), `custom_entangler_layers_normal` and `custom_entangler_layers_uniform`, for easy parameter initialization
- Added tests for the two-parameter initialization functions
- Logged changes in [CHANGELOG.md](http://changelog.md/)

There were a couple of things I wanted to ask specifically that might come up during review:

- Is there anything else I should add to the docstring, under `UseageDetails`? I feel like I should have one or two more examples there.
- Are the tests that I have written extensive enough, or should I write a couple more that explicitly test for **only** edge cases?
- In `test_init.py`, should there be tests that ensure the proper errors are thrown when an invalid pattern is passed as an argument?

Thanks so much, I'm looking forward to hearing your feedback! 🙂

josh146 commented:
@Lucaman99 what is the status of this PR?

josh146 commented:
Hey @Lucaman99, just checking in on the status of this PR. I'm going to close it for now as it appears to be stale. If not feel, free to leave a comment and we can re-open it.
